### PR TITLE
o/snapstate: support user-services when refreshing snaps

### PR DESCRIPTION
**Pull request by Meulengracht**
Implements the remaining functionality to support user-services when refreshing snaps. SnapState now also keeps track of the last disabled services for user-services.

### CODE REVIEW

**Comment by Meulengracht** [comment]
I dropped the spread test I was doing, we were already covering this functionality in the unit tests - and extending them to cover the user services part made sense.

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:3018]: tests never hit here

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:3055]: this comment is not very clear, can we use a single list because hooks cannot target single users via snapctl commands?

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:3219]: I'm not quite sure what is referred here by "one struct"

File: usersession/client/client.go
Comment (pedronis) [inline usersession/client/client.go:336]: does some doc comment need updating in this package?

**Review by pedronis — COMMENTED** [review]
some comments/questions

File: overlord/snapstate/handlers.go
Comment (Meulengracht) [reply overlord/snapstate/handlers.go]: Your comment let me to a missing case of support for user-services. The service tracking code for hooks did not account for user-services - I added this in here, as it is required for completeness in the refresh handling.

File: usersession/client/client.go
Comment (Meulengracht) [reply usersession/client/client.go:336]: Good point! added

**Comment by Meulengracht** [comment]
I split out the wrapper changes: https://github.com/snapcore/snapd/pull/13957

File: overlord/snapstate/handlers.go
Comment (bboozzoo) [inline overlord/snapstate/handlers.go:2980]: could use a comment in case it's not obvious for folks who later look at the code, that the disabled services are collected only from user sessions which are active at the time this code runs

**Review by bboozzoo — DISMISSED** [review]
LGTM, comment tweaks mostly

**Comment by Meulengracht** [comment]
@bboozzoo thanks for the review I'll handle the comments!

I just need some time to think over the strategy here for user-services during refresh. I think I may need to split out a couple of logic steps to provide a better implementation. I have mixed up the purpose of start/stopping and enabling / disabling concepts from a perspective of a per-user state.

File: overlord/servicestate/helpers.go
Comment (Meulengracht) [reply overlord/servicestate/helpers.go:64]: Good point, added better docs in https://github.com/snapcore/snapd/pull/13905/commits/0075433bdfb6cf05d3ac52652b0afa6157523854

File: overlord/servicestate/helpers.go
Comment (Meulengracht) [reply overlord/servicestate/helpers.go:281]: Added!

File: overlord/snapstate/handlers.go
Comment (Meulengracht) [reply overlord/snapstate/handlers.go:3010]: Added!

**Comment by Meulengracht** [comment]
@bboozzoo @ernestl @pedronis I've removed the blocked label, and I've added better explanations for the current approach to kind of back up the reasoning here.

Please see https://github.com/snapcore/snapd/pull/13905/commits/0075433bdfb6cf05d3ac52652b0afa6157523854 for added docs. 
When reviewing this please be *very* critical of code that **should** be better explained, as we will most likely need to come back to this code in the future once we improve our support for per-user state in snapd. I.e it's important that we can understand this again, and unfortunately this state tracking is kinda complex

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:3119]: can we add the InstallMode suffix for these?

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:3135]: this bit of the code is not exercised by tests

**Review by pedronis — COMMENTED** [review]
thanks for the comments, did another pass

File: client/clientutil/session_test.go
Comment (Meulengracht) [reply client/clientutil/session_test.go:40]: good point, added!

File: overlord/servicestate/helpers.go
Comment (Meulengracht) [reply overlord/servicestate/helpers.go]: Doing `if` here was a much nicer solution and made things simpler, thank you for that suggestion

File: overlord/snapstate/handlers.go
Comment (Meulengracht) [reply overlord/snapstate/handlers.go:3142]: is now covered, thank you for spotting this

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:3055]: is this map used in the new code?

File: overlord/snapstate/snapmgr.go
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:307]: does this comment need updates?

**Review by pedronis — COMMENTED** [review]
couple of comments, also it's a bit unclear to me how is --users=all in hooks treated, as it seems this consider only users one-by-one?  but for example --enable --user=all should enable for all users?

**Comment by Meulengracht** [comment]
> couple of comments, also it's a bit unclear to me how is --users=all in hooks treated, as it seems this consider only users one-by-one? but for example --enable --user=all should enable for all users? do we have tests for that scenario?

It's a really good question and highlights that there may be missing some documentation around the solution! If --users=all is set, then there wont be a uid filter, which means it will discover all active user agent sessions and try to affect all of them.

File: overlord/snapstate/snapmgr.go
Comment (Meulengracht) [reply overlord/snapstate/snapmgr.go:322]: I added a small separate comment for the user one, but the comment should be good

File: overlord/snapstate/handlers.go
Comment (Meulengracht) [reply overlord/snapstate/handlers.go]: You are absolutely right - it was a leftover from moving the code!

File: overlord/servicestate/helpers.go
Comment (Meulengracht) [reply overlord/servicestate/helpers.go:97]: added - thank you for pointing this out

File: client/clientutil/session.go
Comment (bboozzoo) [inline client/clientutil/session.go:32]: can we call this code from usersession/client maybe?

File: overlord/servicestate/helpers.go
Comment (bboozzoo) [inline overlord/servicestate/helpers.go:38]: we have similar code in wrappers/services.go, would it make sense to move it to osutil?

File: overlord/snapstate/handlers.go
Comment (bboozzoo) [inline overlord/snapstate/handlers.go:3019]:
```suggestion
// do a best-effort handling of user services - we can only query the user service
```

**Review by bboozzoo — DISMISSED** [review]
Some suggestions

File: client/clientutil/session.go
Comment (Meulengracht) [reply client/clientutil/session.go:32]: good point!

File: overlord/servicestate/helpers.go
Comment (Meulengracht) [reply overlord/servicestate/helpers.go]: moved this

**Comment by Meulengracht** [comment]
I rebased this PR to get the latest master to get test fixes.

File: osutil/user.go
Comment (pedronis) [inline osutil/user.go:418]: this needs a doc comment

File: tests/main/snap-user-services-mixed-refresh-single-disabled/task.yaml
Comment (pedronis) [inline tests/main/snap-user-services-mixed-refresh-single-disabled/task.yaml:100]: we probably can't scrap the global enable/disable but need to have a way to override per user, no? or just error if we get --user and --enable/--disable ?

**Review by pedronis — APPROVED** [review]
thanks, bit confused what is the plan forward but doesn't block the PR as is, the last commits need a re-review by @bboozzoo

File: tests/main/snap-user-services-mixed-refresh-single-disabled/task.yaml
Comment (Meulengracht) [reply tests/main/snap-user-services-mixed-refresh-single-disabled/task.yaml:100]: I will update this comment - this is no longer really valid as we want to maybe support user masking of services instead of preset. I don't think preset is the right way to go anymore

File: osutil/user.go
Comment (Meulengracht) [reply osutil/user.go:418]: good point!

**Review by bboozzoo — APPROVED** [review]
LGTM

**Pull request merged**
